viz/src/bin/draw_occupancy_grid.rs

```diff
@@ -23,85 +25,125 @@ fn draw_frame_marker(window: &mut kiss3d::window::Window, pose: &kuba::Pose3, le
     );
 }
 
+struct GridUpdater {
+    grid_map: std::sync::Mutex<kuba::LidarOccupancyGridMap3>,
+    changed_cells: std::sync::Mutex<Vec<kuba::Cell<na::U3>>>,
+    finished: atomic::AtomicBool,
+}
+
+impl GridUpdater {
+    fn new() -> GridUpdater {
+        GridUpdater {
+            grid_map: std::sync::Mutex::new(kuba::LidarOccupancyGridMap3::default()),
+            changed_cells: std::sync::Mutex::new(vec![]),
+            finished: atomic::AtomicBool::new(true),
+        }
+    }
+}
+
 struct AppState {
     poses: Vec<kuba::Pose3>,
     point_clouds: Vec<kuba::PointCloud3>,
     index: usize,
-    loop_counter: usize,
-    first_loop: bool,
-    grid_map: kuba::LidarOccupancyGridMap3,
+    grid_updater: std::sync::Arc<GridUpdater>,
+    tracked_cells: std::collections::HashMap<kuba::Cell<na::U3>, na::Point3<f32>>,
 }
 
-impl kiss3d::window::State for AppState {
-    fn step(&mut self, window: &mut kiss3d::window::Window) {
-        if self.first_loop || self.loop_counter > 1 {
-            if !self.first_loop {
-                self.loop_counter = 0;
-                self.index = (self.index + 1) % self.point_clouds.len();
-            }
-            self.first_loop = false;
-            let origin = kuba::Point3::from(self.poses[self.index].translation.vector);
-            let start_t = std::time::Instant::now();
-            self.grid_map
-                .integrate_point_cloud(&origin, &self.point_clouds[self.index]);
-            println!("integrate_point_cloud ms: {}", start_t.elapsed().as_millis());
-        }
-        let start_t = std::time::Instant::now();
-        let shape = self.grid_map.grid_map.data.shape();
-        for x in 0..shape[0] {
-            for y in 0..shape[1] {
-                for z in 0..shape[2] {
-                    let cell = kuba::cell3![x as isize, y as isize, z as isize];
-                    if self.grid_map.occupied(&cell) {
-                        let point = self.grid_map.point_from_cell(&cell);
-                        window.draw_point(&point, &na::Point3::new(0.8, 0.0, 0.0));
+impl AppState {
+    fn integrate_next_pointcloud(&mut self) {
+        if self.grid_updater.finished.load(atomic::Ordering::Relaxed) {
+            {
+                let mut changed_cells = self.grid_updater.changed_cells.lock().unwrap();
+                let gridmap = self.grid_updater.grid_map.lock().unwrap();
+
+                for cell in changed_cells.iter() {
+                    let occupied = gridmap.occupied(cell);
+                    if occupied {
```

Comment:
just inline the gridmap.occupied(cell) call in the if statement and remove the occupied intermediate variable.